File: examples/timetabling/moves/SwapFridayWithNonFriday.ts
/**
 * Advanced move operator: Swap Friday class with non-Friday class
 *
 * This operator solves the Friday prayer conflict deadlock by swapping
 * a Friday class with a non-Friday class from the SAME lecturer.
 *
 * Strategy:
 * 1. Find classes on Friday that violate prayer time
 * 2. Find non-Friday classes from the same lecturer
 * 3. Swap their time slots (and rooms if needed)
 * 4. This guarantees no lecturer conflict while fixing Friday issue
 */

import type { MoveGenerator } from '../../../src/index.js';
import type { TimetableState, ScheduleEntry } from '../types/index.js';
import { calculateEndTime, isValidFridayStartTime, canUseExclusiveRoom, isRoomAvailable } from '../utils/index.js';

export class SwapFridayWithNonFriday implements MoveGenerator<TimetableState> {
  name = 'Swap Friday with Non-Friday';

  // Friday prayer time window: 11:40 - 13:10
  private readonly PRAYER_START = 11 * 60 + 40;
  private readonly PRAYER_END = 13 * 60 + 10;

  /**
   * Check if a class overlaps with Friday prayer time
   */
  private overlapsWithPrayerTime(entry: ScheduleEntry): boolean {
    if (entry.timeSlot.day !== 'Friday') {
      return false;
    }

    const [startHour, startMin] = entry.timeSlot.startTime.split(':').map(Number);
    const [endHour, endMin] = entry.timeSlot.endTime.split(':').map(Number);
    const classStart = startHour! * 60 + startMin!;
    const classEnd = endHour! * 60 + endMin!;

    return classStart < this.PRAYER_END && classEnd >= this.PRAYER_START;
  }

  canApply(state: TimetableState): boolean {
    // Only apply if there are Friday violations
    return state.schedule.some(
      (entry) =>
        (entry.timeSlot.day === 'Friday' && !isValidFridayStartTime(entry.timeSlot.startTime)) ||
        this.overlapsWithPrayerTime(entry)
    );
  }

  generate(state: TimetableState, temperature: number): TimetableState {
    // SA engine already clones state, so we work directly on the passed state

    // Find all classes violating Friday constraints
    const fridayViolators = state.schedule.filter(
      (entry) =>
        (entry.timeSlot.day === 'Friday' && !isValidFridayStartTime(entry.timeSlot.startTime)) ||
        this.overlapsWithPrayerTime(entry)
    );

    if (fridayViolators.length === 0) {
      return state;
    }

    // Pick one violating Friday class
    const fridayClass = fridayViolators[Math.floor(Math.random() * fridayViolators.length)]!;

    // Find non-Friday classes from the same lecturer(s)
    const swapCandidates = state.schedule.filter((entry) => {
      // Must be non-Friday
      if (entry.timeSlot.day === 'Friday') return false;

      // Must not be the same class
      if (entry.classId === fridayClass.classId) return false;

      // Must share at least one lecturer
      const sharedLecturer = fridayClass.lecturers.some((lec) => entry.lecturers.includes(lec));
      if (!sharedLecturer) return false;

      return true;
    });

    if (swapCandidates.length === 0) {
      return state; // No swap candidates available
    }

    // Pick a random swap candidate
    const nonFridayClass = swapCandidates[Math.floor(Math.random() * swapCandidates.length)]!;

    // Check if swap is feasible (room compatibility)
    const fridayClassCanUseNonFridayRoom = this.canSwapRoom(
      state,
      fridayClass,
      nonFridayClass.room,
      nonFridayClass.timeSlot.day
    );

    const nonFridayClassCanUseFridayRoom = this.canSwapRoom(
      state,
      nonFridayClass,
      fridayClass.room,
      fridayClass.timeSlot.day
    );

    // If rooms are compatible, do a simple swap
    if (fridayClassCanUseNonFridayRoom && nonFridayClassCanUseFridayRoom) {
      // Swap time slots
      const tempTimeSlot = { ...fridayClass.timeSlot };
      const tempRoom = fridayClass.room;
      const tempPrayerTime = fridayClass.prayerTimeAdded;

      fridayClass.timeSlot = { ...nonFridayClass.timeSlot };
      fridayClass.room = nonFridayClass.room;
      fridayClass.prayerTimeAdded = nonFridayClass.prayerTimeAdded;

      nonFridayClass.timeSlot = tempTimeSlot;
      nonFridayClass.room = tempRoom;
      nonFridayClass.prayerTimeAdded = tempPrayerTime;

      return state;
    }

    // If rooms incompatible, try swapping time slots and finding new rooms
    // Swap time slots only
    const tempTimeSlot = { ...fridayClass.timeSlot };
    fridayClass.timeSlot = { ...nonFridayClass.timeSlot };
    nonFridayClass.timeSlot = tempTimeSlot;

    // Try to find suitable rooms for both classes at their new time slots
    const fridayClassNewRoom = this.findSuitableRoom(state, fridayClass);
    const nonFridayClassNewRoom = this.findSuitableRoom(state, nonFridayClass);

    if (fridayClassNewRoom && nonFridayClassNewRoom) {
      fridayClass.room = fridayClassNewRoom;
      nonFridayClass.room = nonFridayClassNewRoom;

      // Recalculate prayer times
      const calc1 = calculateEndTime(fridayClass.timeSlot.startTime, fridayClass.sks, fridayClass.timeSlot.day);
      const calc2 = calculateEndTime(
        nonFridayClass.timeSlot.startTime,
        nonFridayClass.sks,
        nonFridayClass.timeSlot.day
      );

      fridayClass.prayerTimeAdded = calc1.prayerTimeAdded;
      nonFridayClass.prayerTimeAdded = calc2.prayerTimeAdded;

      fridayClass.timeSlot.endTime = calc1.endTime;
      nonFridayClass.timeSlot.endTime = calc2.endTime;

      return state;
    }

    // Swap failed - revert and return original
    // Need to swap back since we modified in place
    const revertTimeSlot = { ...fridayClass.timeSlot };
    fridayClass.timeSlot = { ...nonFridayClass.timeSlot };
    nonFridayClass.timeSlot = revertTimeSlot;
    
    return state;
  }

  /**
   * Check if a class can use a specific room at a specific day
   */
  private canSwapRoom(state: TimetableState, entry: ScheduleEntry, newRoom: string, newDay: string): boolean {
    // Check exclusive room constraints
    if (!canUseExclusiveRoom(newRoom, entry.className, entry.prodi)) {
      return false;
    }

    // Check capacity
    const room = state.rooms.find((r) => r.Code === newRoom);
    if (!room || room.Capacity < entry.participants) {
      return false;
    }

    // Check lab requirement
    if (entry.needsLab && !room.Type.toLowerCase().includes('lab')) {
      return false;
    }

    return true;
  }

  /**
   * Find a suitable room for a class at its current time slot
   */
  private findSuitableRoom(state: TimetableState, entry: ScheduleEntry): string | null {
    const otherSchedule = state.schedule.filter((e) => e.classId !== entry.classId);

    const suitableRooms = state.rooms.filter((room) => {
      // Check capacity
      if (room.Capacity < entry.participants) return false;

      // Check exclusive room
      if (!canUseExclusiveRoom(room.Code, entry.className, entry.prodi)) return false;

      // Check lab requirement
      if (entry.needsLab && !room.Type.toLowerCase().includes('lab')) return false;

      // Check availability
      if (!isRoomAvailable(otherSchedule, room.Code, entry.timeSlot, entry.sks)) return false;

      return true;
    });

    if (suitableRooms.length === 0) return null;

    // Return smallest suitable room (to save larger rooms)
    suitableRooms.sort((a, b) => a.Capacity - b.Capacity);
    return suitableRooms[0].Code;
  }
}

File: python/src/timetable_sa/examples/timetabling/moves/swap_friday_with_non_friday.py
"""Move generator to swap Friday classes with non-Friday days."""

import random
from typing import Optional
from timetable_sa.core.interfaces import MoveGenerator
from timetable_sa.examples.timetabling.domain_types.state import TimetableState, ScheduleEntry
from timetable_sa.examples.timetabling.domain_types.domain import TimeSlot
from timetable_sa.examples.timetabling.utils.initial_solution import clone_state


class SwapFridayWithNonFriday(MoveGenerator):
    """Move generator that moves Friday classes to non-Friday days.
    
    This is a targeted move generator for fixing Friday-related violations.
    """
    
    name: str = "Swap Friday With Non Friday"
    
    def __init__(self, target_days: Optional[list[str]] = None):
        """Initialize swap Friday with non-Friday move generator.
        
        Args:
            target_days: List of non-Friday days to consider (default: all except Friday)
        """
        self.target_days = target_days or [
            "Monday", "Tuesday", "Wednesday", "Thursday"
        ]
    
    def can_apply(self, state: TimetableState) -> bool:
        """Check if this move can be applied."""
        return any(
            entry.time_slot.day == "Friday"
            for entry in state.schedule
        )
    
    def generate(self, state: TimetableState, temperature: float) -> Optional[TimetableState]:
        """Generate a new state with a Friday class moved to a non-Friday day.
        
        Args:
            state: Current timetable state
            temperature: Current annealing temperature
            
        Returns:
            New state with Friday class moved, or None if cannot apply
        """
        if not self.can_apply(state):
            return None
        
        friday_entries = [
            entry for entry in state.schedule
            if entry.time_slot.day == "Friday"
        ]
        
        entry = random.choice(friday_entries)
        
        available_slots = []
        for day in self.target_days:
            for period in range(1, 13):
                new_time_slot = next(
                    (slot for slot in state.available_time_slots 
                     if slot.day == day and slot.period == period),
                    None
                )
                if new_time_slot:
                    is_available = not any(
                        e.room == entry.room and
                        e.time_slot.day == day and
                        e.time_slot.period == period
                        for e in state.schedule
                    )
                    
                    lecturer_available = not any(
                        e.lecturers[0] if e.lecturers else None == entry.lecturers[0] if entry.lecturers else None and
                        e.time_slot.day == day and
                        e.time_slot.period == period
                        for e in state.schedule
                    )
                    
                    prodi_available = not any(
                        e.prodi == entry.prodi and
                        e.time_slot.day == day and
                        e.time_slot.period == period
                        for e in state.schedule
                    )
                    
                    if is_available and lecturer_available and prodi_available:
                        available_slots.append(new_time_slot)
        
        if not available_slots:
            return None
        
        new_time_slot = random.choice(available_slots)
        new_state = clone_state(state)
        
        for i, e in enumerate(new_state.schedule):
            if e.class_id == entry.class_id and e.kelas == entry.kelas:
                from timetable_sa.examples.timetabling.utils.time import calculate_end_time
                end_time, _ = calculate_end_time(new_time_slot.start_time, e.sks, new_time_slot.day)
                
                new_state.schedule[i] = ScheduleEntry(
                    class_id=e.class_id,
                    class_name=e.class_name,
                    kelas=e.kelas,
                    prodi=e.prodi,
                    lecturers=e.lecturers,
                    room=e.room,
                    time_slot=TimeSlot(
                        day=new_time_slot.day,
                        start_time=new_time_slot.start_time,
                        end_time=end_time,
                        period=new_time_slot.period
                    ),
                    sks=e.sks,
                    needs_lab=e.needs_lab,
                    participants=e.participants,
                    class_type=e.class_type,
                    prayer_time_added=e.prayer_time_added,
                    is_overflow_to_lab=e.is_overflow_to_lab,
                )
                break
        
        return new_state
